File: algorithms/String/String-compression/string-compression.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func main() {
	fmt.Println(CompressString("aabbccccccdddddddddd"))
}

func CompressString(input string) string {

	if len(input) == 0 {
		return " "
	}

	var builder strings.Builder

	currentChar := input[0]
	charCount := 1

	for i := 1; i < len(input); i++ {
		if input[i] == currentChar {
			charCount++
		} else {
			writeCharAndCount(&builder, currentChar, charCount)
			currentChar = input[i]
			charCount = 1
		}
	}
	// we have to write this funtion outside of loop because loop ends before writing in the builder.
	writeCharAndCount(&builder, currentChar, charCount)

	return builder.String()

}
func writeCharAndCount(builder *strings.Builder, char byte, count int) {
	builder.WriteByte(char)
	if count > 1 {
		builder.WriteString(strconv.Itoa(count))
	}
}

File: algorithms/String/String-compression/string-compression.js
function compressString(input) {
  if (input.length === 0) {
    return " ";
  }

  let result = "";
  let currentChar = input[0];
  let charCount = 1;

  for (let i = 1; i < input.length; i++) {
    if (input[i] === currentChar) {
      charCount++;
    } else {
      result += writeCharAndCount(currentChar, charCount);
      currentChar = input[i];
      charCount = 1;
    }
  }
  // Write the last character and its count because last character not have been written.
  result += writeCharAndCount(currentChar, charCount);

  return result;
}

function writeCharAndCount(char, count) {
  return char + (count > 1 ? count : "");
}

// Example usage:
console.log(compressString("aabbccccccdddddddddd")); // Output: "a2b2c6d10"
